File: extract_brokers.py
import pandas as pd
import json
import os

# قراءة ملف الإكسل
excel_path = "/home/ubuntu/upload/داتا الوسطاء العقاريين بالمملكة .xlsx"
output_path = "/home/ubuntu/wasset_website/brokers_data.js"

try:
    # قراءة ملف الإكسل
    df = pd.read_excel(excel_path)
    
    # طباعة أسماء الأعمدة للتحقق
    print("أسماء الأعمدة في الملف:")
    print(df.columns.tolist())
    
    # طباعة عدد الصفوف
    print(f"عدد الوسطاء في الملف: {len(df)}")
    
    # طباعة بعض البيانات للتحقق
    print("\nعينة من البيانات:")
    print(df.head())
    
    # تحضير قائمة الوسطاء
    brokers = []
    
    # استخراج البيانات من كل صف
    for index, row in df.iterrows():
        broker = {}
        
        # محاولة استخراج البيانات الأساسية
        try:
            # التحقق من نوع الوسيط (فرد أو منشأة)
            broker_type = None
            broker_name = None
            phone = None
            
            # محاولة استخراج الاسم
            for col in df.columns:
                if 'اسم' in col.lower() and pd.notna(row[col]):
                    broker_name = str(row[col]).strip()
                    break
            
            # محاولة استخراج رقم الجوال
            for col in df.columns:
                if 'جوال' in col.lower() or 'هاتف' in col.lower() or 'رقم' in col.lower():
                    if pd.notna(row[col]):
                        phone = str(row[col]).strip()
                        # تنظيف رقم الهاتف
                        phone = ''.join(filter(str.isdigit, phone))
                        break
            
            # محاولة تحديد نوع الوسيط
            for col in df.columns:
                if 'نوع' in col.lower() and pd.notna(row[col]):
                    type_value = str(row[col]).strip().lower()
                    if 'فرد' in type_value or 'شخص' in type_value:
                        broker_type = 'individual'
                    elif 'منشأة' in type_value or 'شركة' in type_value or 'مؤسسة' in type_value:
                        broker_type = 'company'
                    break
            
            # إذا لم يتم تحديد النوع، نحاول تخمينه من الاسم
            if broker_type is None and broker_name:
                if ' ' in broker_name and len(broker_name.split()) <= 4:
                    broker_type = 'individual'  # نفترض أنه فرد إذا كان الاسم يتكون من كلمتين إلى أربع كلمات
                else:
                    broker_type = 'company'  # نفترض أنه منشأة في الحالات الأخرى
            
            # إذا تم العثور على الاسم ورقم الهاتف
            if broker_name and phone:
                broker['name'] = broker_name
                broker['phone'] = phone
                broker['type'] = broker_type
                
                # تحديد اسم المستخدم وكلمة المرور حسب النوع
                if broker_type == 'individual':
                    # للأفراد: الاسم الأول والأخير
                    name_parts = broker_name.split()
                    if len(name_parts) >= 2:
                        broker['username'] = name_parts[0] + ' ' + name_parts[-1]
                    else:
                        broker['username'] = broker_name
                else:
                    # للمنشآت: اسم المنشأة كاملاً
                    broker['username'] = broker_name
                
                # كلمة المرور هي رقم الجوال في كلتا الحالتين
                broker['password'] = phone
                
                # إضافة الوسيط إلى القائمة
                brokers.append(broker)
        
        except Exception as e:
            print(f"خطأ في معالجة الصف {index}: {e}")
    
    # طباعة عدد الوسطاء المستخرجين
    print(f"\nتم استخراج بيانات {len(brokers)} وسيط بنجاح")
    
    # تحويل البيانات إلى تنسيق JavaScript
    js_content = f"const brokersData = {json.dumps(brokers, ensure_ascii=False, indent=2)};"
    
    # إنشاء مجلد الموقع إذا لم يكن موجوداً
    os.makedirs(os.path.dirname(output_path), exist_ok=True)
    
    # حفظ البيانات في ملف JavaScript
    with open(output_path, 'w', encoding='utf-8') as f:
        f.write(js_content)
    
    print(f"\nتم حفظ بيانات الوسطاء في الملف: {output_path}")
    
except Exception as e:
    print(f"حدث خطأ أثناء معالجة ملف الإكسل: {e}")

File: extract_brokers.js
// سكريبت معالجة بيانات الوسطاء العقاريين من ملفات Excel إلى JSON

const fs = require('fs');
const xlsx = require('xlsx');
const path = require('path');

// تحديد مسارات الملفات
const excelFilePath = path.join(__dirname, 'الوسطاء العقاريين بالمملكة.xlsx');
const outputJsonPath = path.join(__dirname, 'brokers_data.json');
const outputJsPath = path.join(__dirname, 'brokers_data.js');

// دالة لتحويل ملف Excel إلى JSON
function convertExcelToJson() {
    try {
        console.log('بدء معالجة ملف Excel...');
        
        // قراءة ملف Excel
        const workbook = xlsx.readFile(excelFilePath);
        const sheetName = workbook.SheetNames[0];
        const worksheet = workbook.Sheets[sheetName];
        
        // تحويل ورقة العمل إلى مصفوفة من الكائنات
        const rawData = xlsx.utils.sheet_to_json(worksheet);
        
        console.log(`تم قراءة ${rawData.length} سجل من ملف Excel.`);
        
        // معالجة البيانات وتنظيفها
        const brokers = processData(rawData);
        
        // حفظ البيانات كملف JSON
        fs.writeFileSync(outputJsonPath, JSON.stringify(brokers, null, 2), 'utf8');
        console.log(`تم حفظ البيانات في ${outputJsonPath}`);
        
        // إنشاء ملف JavaScript
        const jsContent = `// بيانات الوسطاء العقاريين المرخصين
const brokersData = ${JSON.stringify(brokers, null, 2)};

// دالة للبحث عن وسيط باستخدام اسم المستخدم وكلمة المرور
function validateBroker(username, password) {
    return brokersData.find(broker => {
        // للوسيط الفرد: اسم المستخدم هو الاسم الأول والأخير، وكلمة المرور هي رقم الجوال
        if (broker.type === 'individual') {
            return (broker.fullName === username && broker.mobile === password);
        }
        // للمنشأة العقارية: اسم المستخدم هو اسم المنشأة، وكلمة المرور هي رقم الجوال
        else if (broker.type === 'company') {
            return (broker.companyName === username && broker.mobile === password);
        }
        return false;
    });
}

// دالة للحصول على جميع الوسطاء في منطقة معينة
function getBrokersByRegion(region) {
    return brokersData.filter(broker => broker.region === region);
}

// دالة للحصول على الوسطاء حسب نوعهم (فرد أو منشأة)
function getBrokersByType(type) {
    return brokersData.filter(broker => broker.type === type);
}

// دالة للبحث عن الوسطاء باستخدام كلمة مفتاحية
function searchBrokers(keyword) {
    const searchTerm = keyword.toLowerCase();
    return brokersData.filter(broker => {
        return (
            (broker.fullName && broker.fullName.toLowerCase().includes(searchTerm)) ||
            (broker.companyName && broker.companyName.toLowerCase().includes(searchTerm)) ||
            (broker.region && broker.region.toLowerCase().includes(searchTerm)) ||
            (broker.city && broker.city.toLowerCase().includes(searchTerm))
        );
    });
}`;
        
        fs.writeFileSync(outputJsPath, jsContent, 'utf8');
        console.log(`تم إنشاء ملف JavaScript في ${outputJsPath}`);
        
        return {
            success: true,
            message: 'تم معالجة البيانات بنجاح',
            count: brokers.length
        };
    } catch (error) {
        console.error('حدث خطأ أثناء معالجة الملف:', error);
        return {
            success: false,
            message: `حدث خطأ أثناء المعالجة: ${error.message}`
        };
    }
}

// دالة لمعالجة وتنظيف البيانات
function processData(rawData) {
    console.log('بدء معالجة وتنظيف البيانات...');
    
    // تحويل أسماء الحقول إلى تنسيق موحد
    const brokers = rawData.map((record, index) => {
        // تحديد نوع الوسيط (فرد أو منشأة)
        const isCompany = record['نوع الوسيط'] === 'منشأة عقارية' || 
                         record['نوع الترخيص'] === 'منشأة' ||
                         record['اسم المنشأة'] !== undefined;
        
        // إنشاء كائن موحد لكل وسيط
        const broker = {
            id: index + 1,
            type: isCompany ? 'company' : 'individual',
            licenseNumber: record['رقم الترخيص'] || record['رقم رخصة فال'] || '',
            licenseDate: record['تاريخ الترخيص'] || '',
            licenseExpiry: record['تاريخ انتهاء الترخيص'] || '',
            region: record['المنطقة'] || '',
            city: record['المدينة'] || '',
            mobile: record['رقم الجوال'] || record['الجوال'] || '',
            email: record['البريد الإلكتروني'] || '',
            status: record['حالة الترخيص'] || 'نشط',
            points: Math.floor(Math.random() * 1000), // نقاط عشوائية للتجربة
            deals: Math.floor(Math.random() * 50), // عدد الصفقات عشوائي للتجربة
            rating: (Math.random() * 2 + 3).toFixed(1), // تقييم عشوائي بين 3 و 5
            joinDate: generateRandomDate(2020, 2023)
        };
        
        // إضافة حقول خاصة بالوسيط الفرد
        if (!isCompany) {
            broker.fullName = record['اسم الوسيط'] || record['الاسم'] || '';
            broker.firstName = broker.fullName.split(' ')[0] || '';
            broker.lastName = broker.fullName.split(' ').slice(-1)[0] || '';
            broker.idNumber = record['رقم الهوية'] || '';
        } 
        // إضافة حقول خاصة بالمنشأة العقارية
        else {
            broker.companyName = record['اسم المنشأة'] || record['اسم الشركة'] || '';
            broker.commercialRegister = record['رقم السجل التجاري'] || '';
            broker.managerName = record['اسم المدير'] || '';
        }
        
        return broker;
    });
    
    console.log(`تم معالجة ${brokers.length} سجل.`);
    
    // إزالة السجلات المكررة
    const uniqueBrokers = removeDuplicates(brokers);
    console.log(`تم إزالة السجلات المكررة. العدد النهائي: ${uniqueBrokers.length}`);
    
    return uniqueBrokers;
}

// دالة لإزالة السجلات المكررة
function removeDuplicates(brokers) {
    const uniqueMap = new Map();
    
    brokers.forEach(broker => {
        const key = broker.licenseNumber || broker.mobile;
        if (!uniqueMap.has(key) || broker.licenseNumber) {
            uniqueMap.set(key, broker);
        }
    });
    
    return Array.from(uniqueMap.values());
}

// دالة لإنشاء تاريخ عشوائي
function generateRandomDate(startYear, endYear) {
    const year = Math.floor(Math.random() * (endYear - startYear + 1)) + startYear;
    const month = Math.floor(Math.random() * 12) + 1;
    const day = Math.floor(Math.random() * 28) + 1;
    
    return `${year}-${month.toString().padStart(2, '0')}-${day.toString().padStart(2, '0')}`;
}

// تنفيذ التحويل
const result = convertExcelToJson();
console.log(result);

module.exports = {
    convertExcelToJson
};
